### ISSUE TITLE
To have single cuda context across multiple processes

### ISSUE THREAD

**Issue by hemantranvir** [pytorch/pytorch#42080]
Presently, as far I have digged around, pytorch doesn't seem to use single cuda context across multiple processes. If training the network in multiple processes the 557 MB per cuda context in each process is quite expensive. It would be nice to enable pytorch to use single cuda context.

FYI: 
- The multiple process training requirement could be mitigated using `torch.multiprocessing` but it would be good to have it for legacy processes too.
- I tried using cuda Multi Process Service (MPS) which should by default use single cuda context no matter where you are spawning the different processes. But even then for every new process spawned the GPU memory used is almost equivalent.

It would be nice if someone could point me in the right direction or in the source code of pytorch, where torch spawns cuda contexts and I could change the context creation to get the present cuda context and use it if available. (given it would be minor change)

**Comment by ngimel** [comment]
The correct solution here is to use MPS, making new process use the current context inside pytorch would be an unexpected behavior.

**Comment by hemantranvir** [comment]
@ngimel  Thanks for your response! Some more pointers:

For the MPS:
I started the MPS using:
```
$ sudo nvidia-smi -i 0 -c EXCLUSIVE_PROCESS 
$ sudo nvidia-cuda-mps-control -d
```

And then started two similar python processes in different shells/processes using:
```
$ python
Python 3.7.8 (default, Jun 29 2020, 05:44:46) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.randn(1,device=0)
tensor([-1.2574], device='cuda:0')
>>> 
```

When checking `nvidia-smi` output:
**Without MPS**
```
$ nvidia-smi
Mon Jul 27 06:53:59 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.33.01    Driver Version: 440.33.01    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 108...  Off  | 00000000:02:00.0 Off |                  N/A |
|  0%   44C    P2    68W / 270W |   1120MiB / 11175MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0     12608      C   python                                       555MiB |
|    0     12802      C   python                                       555MiB |
+-----------------------------------------------------------------------------+
$
```

**With MPS**
```
$ nvidia-smi
Mon Jul 27 06:29:45 2020       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.33.01    Driver Version: 440.33.01    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 108...  Off  | 00000000:02:00.0 Off |                  N/A |
|  0%   33C    P2    66W / 270W |   1229MiB / 11175MiB |      0%   E. Process |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|    0      1674      G   /usr/lib/xorg/Xorg                            63MiB |
|    0      1800      G   /usr/bin/gnome-shell                          89MiB |
|    0     30642      C   nvidia-cuda-mps-server                      1071MiB |
+-----------------------------------------------------------------------------+
$ 
```

When MPS is on, it should use just one single cuda context, thus the GPU memory consumption should have been almost half of 1071MiB. May I know why is that the case or how can I make it use single cuda context?

**Comment by ngimel** [comment]
cc @ptrblck, do you know why there's no context sharing with mps?

**Comment by ptrblck** [comment]
I don't know, but I can reproduce it locally.
I'll try to dig into it and ask our CUDA team what might be the reason.

**Comment by hemantranvir** [comment]
@ptrblck Thank you for the response!
Look forward to it.

**Comment by hemantranvir** [comment]
@ptrblck Howdy! Is there any update on this. Thanks!

**Comment by mcarilli** [comment]
Your use case is unclear.  Do you
1. want each process to use its own device (or device subset) and not create spurious contexts on other devices, or 
2. want multiple processes to share one device? 

If 1, each Pytorch process lazily creates a context on a particular device only when it attempts to use that device.  If you want to ensure each process creates a context on only one device (or a subset of devices) all you have to do is make sure that process doesn't touch any other devices.  Calling torch.cuda.set_device() early in each process can help ensure this.

If 2, we should double check whether MPS allows context sharing across processes, or if it multiplexes calls from different processes with different contexts so they can run on the GPU simultaneously.  The latter would not provide memory savings but still exposes parallelism across processes sharing the same device

**Comment by hemantranvir** [comment]
@mcarilli Thanks for the reply. The use case is number 2.
Referring to the following slides (https://on-demand.gputechconf.com/gtc/2015/presentation/S5584-Priyanka-Sah.pdf), which I believe is a presentation from the gpu tech conference, if you refer slide 4:
```
CUDA MPS is a feature that allows multiple CUDA processes to share a single GPU context. each process receive some subset of the available connections to that GPU.
```

Moreover, referring slide 12 and 21 , seems like MPS does allow sharing one single cuda context across processes.
I believe there are primary contexts and other user-created contexts, that can be created using cuda api. If torch uses primary context i.e does not create a custom cuda context then the context sharing should happen by default.

Looking at my previous findings though, it might suggest that it does create custom cuda context? It is just my suspicion, would be grateful if you could help me in this regard.
Thanks!

**Comment by ZinuoCai** [comment]
Any progress?

**Comment by maaft** [comment]
Also interested!

**Comment by sharvil** [comment]
Wouldn't `CUDACachingAllocator` prevent multiple PyTorch processes from training on the same GPU (in general) even if MPS allowed sharing CUDA contexts?

Imagine two training processes sharing a GPU that have a peak GPU memory use of 30% each. It's possible for the caching allocator on process 1 to take ownership of 90% of GPU memory. Process 2 has no knowledge of process 1's allocator state. When process 2 attempts to allocate the 30% it needs, it won't be able to flush process 1's cache and will OOM.

**Comment by maaft** [comment]
Not sure if this is a real issue in practice.

We do exactly this in our AIaaS prod environment and had no issues since 3 years.

**Comment by sharvil** [comment]
It's entirely possible that some workloads don't encounter this problem in practice, but that doesn't necessarily mean all valid workloads will succeed. I consistently run into OOMs when training multiple models on a single GPU even though the sum of their peak mem use is a few GB less than total available GPU memory.

**Comment by fxmarty** [comment]
@d4l3k From what I understood from your presentation yesterday at PyTorch Conference, in the inference-only case, torch::deploy/multipy would somewhat mitigate the issue in case I have a GPU, and run several models on it? Could it allow to share the CUDA context?

**Comment by d4l3k** [comment]
@fxmarty yes exactly, torch::deploy is all in a single process so you can share the cuda context across multiple interpreters/GPUs

Let me know if you need any help with that. You also could use torch::deploy for training but you'd have to write some custom glue logic currently

**Comment by venkataganti** [comment]
fdfsfs
